snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+
+	if ratio > stakeThresholdForStragglerSuspicion {
+		sd.log.Trace("We are straggling behind", zap.Float64("ratio", ratio))
+		return true
+	}
+
+	sd.log.Trace("Nodes ahead of us:", zap.Float64("ratio", ratio))
+
+	return false
+}
+
+func (sd *stragglerDetector) validateNetInfo(netInfo netInfo) bool {
+	if netInfo.connStakePercent < sd.minConfirmationThreshold {
+		// We don't know for sure whether we're behind or not.
+		// Even if we're behind, it's pointless to act before we have established
+		// connectivity with enough validators.
+		sd.log.Verbo("not enough connected stake to determine network info", zap.Float64("ratio", netInfo.connStakePercent))
+		return false
+	}
+
+	if netInfo.totalValidatorWeight == 0 {
+		sd.log.Trace("Connected to zero weight")
+		return false
+	}
+
+	totalKnownLastBlockStakePercent := float64(netInfo.totalWeightWeKnowItsLastAcceptedBlock) / float64(netInfo.totalValidatorWeight)
+	stakeAheadOfUs := float64(netInfo.totalPendingStake) / float64(netInfo.totalValidatorWeight)
+
+	// Ensure we have collected last accepted blocks for at least 80% (or so) stake of the total weight we are connected to.
+	if totalKnownLastBlockStakePercent < minimumStakeThresholdRequiredForNetworkInfo {
+		sd.log.Trace("Not collected enough information about last accepted blocks for the validators we are connected to",
+			zap.Float64("ratio", totalKnownLastBlockStakePercent))
+		return false
+	}
+
+	if stakeAheadOfUs < knownStakeThresholdRequiredForAnalysis {
+		sd.log.Trace("Most stake we're connected to has the same height as we do",
+			zap.Float64("ratio", stakeAheadOfUs))
+		return false
+	}
+
+	return true
+}
+
+func (sd *stragglerDetector) getNetworkSnapshot() (snapshot, bool) {
+	ourLastAcceptedBlock := sd.lastAccepted()
+
+	netInfo, err := sd.getNetworkInfo(ourLastAcceptedBlock)
+	if err != nil {
+		return snapshot{}, false
+	}
+
+	if !sd.validateNetInfo(netInfo) {
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		nodeWeightsToBlocks:  netInfo.nodeWeightToLastAccepted,
+		totalValidatorWeight: netInfo.totalValidatorWeight,
+	}
+
+	if sd.haveWeFailedCatchingUp(snap) {
+		return snap, true
+	}
+
+	return snapshot{}, false
+}
+
+type netInfo struct {
+	connStakePercent                      float64
+	totalPendingStake                     uint64
+	totalValidatorWeight                  uint64
+	totalWeightWeKnowItsLastAcceptedBlock uint64
+	nodeWeightToLastAccepted              nodeWeightsToBlocks
+}
+
+func (sd *stragglerDetector) getNetworkInfo(ourLastAcceptedBlock ids.ID) (netInfo, error) {
+	var res netInfo
+
+	res.connStakePercent = sd.connectedPercent()
+
+	validators := nodeWeights(sd.connectedValidators().List())
+	nodeWeightToLastAccepted := make(nodeWeightsToBlocks, len(validators))
+
+	for _, vdr := range validators {
+		lastAccepted, ok := sd.lastAcceptedByNodeID(vdr.Node)
+		if !ok {
+			continue
+		}
+		nodeWeightToLastAccepted[vdr] = lastAccepted
+	}
+
+	totalValidatorWeight, err := validators.totalWeight()
```

Comment:
part of me wants to say we should just use `ConnectedWeight` but seems it can be racy.

Reply:
using connected weight would indeed be not memory coherent. We cannot use it, unlike the connected percent.